pkg/querier/querier.go

```diff
@@ -212,3 +226,101 @@ func mergePair(s1, s2 []string) []string {
 	}
 	return result
 }
+
+// Tail keeps getting matching logs from all ingesters for given query
+func (q *Querier) Tail(ctx context.Context, req *logproto.TailRequest) (*Tailer, error) {
+	clients, err := q.forAllIngesters(func(client logproto.QuerierClient) (interface{}, error) {
+		return client.Tail(ctx, req)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	tailClients := make(map[string]logproto.Querier_TailClient)
+	for i := range clients {
+		tailClients[clients[i].addr] = clients[i].response.(logproto.Querier_TailClient)
+	}
+
+	return newTailer(time.Duration(req.DelayFor)*time.Second, tailClients, func(from, to time.Time, labels string) (iterator iter.EntryIterator, e error) {
+		return q.queryDroppedStreams(ctx, req, from, to, labels)
+	}, func(connectedIngestersAddr []string) (map[string]logproto.Querier_TailClient, error) {
+		return q.tailDisconnectedIngesteres(ctx, req, connectedIngestersAddr)
+	}), nil
+}
+
+// passed to tailer for querying dropped streams
+func (q *Querier) queryDroppedStreams(ctx context.Context, req *logproto.TailRequest, start, end time.Time, labels string) (iter.EntryIterator, error) {
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	key := token_util.TokenFor(userID, labels)
+	replicationSet, err := q.ring.Get(key, ring.Read)
+	if err != nil {
+		return nil, err
+	}
+
+	query := logproto.QueryRequest{
+		Direction: logproto.FORWARD,
+		Start:     start,
+		End:       end,
+		Query:     req.Query,
+		Regex:     req.Regex,
+		Limit:     10000,
+	}
+
+	clients, err := q.forGivenIngesters(replicationSet, func(client logproto.QuerierClient) (interface{}, error) {
+		return client.Query(ctx, &query)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	ingesterIterators := make([]iter.EntryIterator, len(clients))
+	for i := range clients {
+		ingesterIterators[i] = iter.NewQueryClientIterator(clients[i].response.(logproto.Querier_QueryClient), query.Direction)
+	}
+
+	chunkStoreIterators, err := q.queryStore(ctx, &query)
+	if err != nil {
+		return nil, err
+	}
+
+	iterators := append(ingesterIterators, chunkStoreIterators)
+	return iter.NewHeapIterator(iterators, query.Direction), nil
+}
+
+// passed to tailer for (re)connecting to new or disconnected ingesters
+func (q *Querier) tailDisconnectedIngesteres(ctx context.Context, req *logproto.TailRequest, connectedIngestersAddr []string) (map[string]logproto.Querier_TailClient, error) {
```

Comment:
tailDisconnectedIngesteres - > tailDisconnectedIngesters

Reply:
Renamed it. Thanks!